<table style="width :100%;">
    <td style="width : 7.5%;">
        <img style="float: left;
                width: 75px;"
         src="https://upload.wikimedia.org/wikipedia/fr/d/d9/Logo_T%C3%A9l%C3%A9com_ParisTech.svg" />
    </td>
    <td style="width : 15%;">
        TELECOM PARIS<br />2021/2022<br /><br /> MDI720<br />Mme Anne Sabourin
    </td>
    <td>
        <h1><center>STATISTIQUES<br /><br />TP2 : Régression Linéaire</center></h1>
    </td>
    <td style="width : 12.5%;">
        Victor LEDEZ <br />Date TP : 2021/10/29<br /><br /> Date CR : 2021/11/10<br /> Version : 1
    </td>
</table>

Vous devez déposer votre fichier, sous format nomdefamille.ipynb, sur le site ecampus avant le 10/11, 23h59 sur ecampus dans l’espace de rendu dédié. Pour que la correction par les pairs soit possible, aucun retard ne sera toléré (en l’absence de rendu à la date limite la note attribuée sera zéro). La correction par les pairs se fera selon les mêmes modalités qu’au premier TP noté. Ne tenez pas compte du barême pour la correction, notez toujours sur 2 points chaque question, la pondération s’effectue automatiquement dans la feuille excel. Date limite de rendu de la correction : le 21/11 23h59.


<hr/>

> Nous travaillons sur la base de données diabetes de python. La base initiale comporte n “ 442 patients et p “ 10 covariables. La variable Y à expliquer est un score correspondant à l’évolution de la maladie. Pour s’amuser, un robot malicieux a contaminé le jeu de données en y ajoutant 150 variables explicatives inappropriées. Ensuite, non-content d’avoir déjà perverti notre jeu de données, il a volontairement mélangé les variables entre elles de façon aléatoire. Bien entendu le robot a ensuite pris soin d’effacer toute trace de son acte crapuleux si bien que nous ne connaissons pas les variables pertinentes.La nouvelle base de données notée X comporte n “ 442 patients et p “ 160 covariables. Saurez-vous déjouer les plans de ce robot farceur et retrouver les variables pertinentes ?

<hr/>

In [8]:
import pandas as pd


<hr>

## Question 1

<hr>

> Importer la base de données data_contamine.csv disponible depuis le lien https://github.com/anassag/TP-reduction-dimension/blob/main/data_contamine.csv. La dernière colonne est la variable à expliquer. Les autres colonnes sont les variables explicatives. Vérifiez le nombre de variables explicatives et le nombre d’observations. (1 point)

In [9]:

data = pd.read_csv('data_contamine.csv')
data.head()


2.237733528583032705e-01  2.407005121478057230e+00  \
0                 -0.515789                 -0.799958   
1                  0.023454                  0.048646   
2                  0.892292                 -0.494714   
3                 -1.021634                 -1.805761   
4                 -1.288486                 -2.337547   

   -5.360732885543344484e-01  -1.336625915483726335e+00  \
0                   1.582195                  -0.008126   
1                   0.224159                   0.655717   
2                  -1.062153                   0.736429   
3                   0.082164                   0.387965   
4                  -0.186132                  -1.230419   

   -9.794715446704517492e-02  1.549540399810333424e+00  \
0                   0.414271                  1.363709   
1                  -0.176065                  1.605403   
2                  -0.118200                 -1.078412   
3                  -0.838219                  1.596253   
4                  -0.063607                 -0.928346   

   1.590890177396816707e+00  1.190101277089628207e+00  \
0                 -0.098582                 -0.302917   
1                  1.638630                 -1.240650   
2                 -0.454316                 -0.378080   
3                  0.037945                 -1.012175   
4                 -1.138335                 -1.180762   

   -1.181935530900874776e+00  2.731035665376587929e+00  ...  \
0                  -0.544602                 -0.241896  ...   
1                   2.100396                 -1.031570  ...   
2                   2.679422                 -0.051751  ...   
3                   1.632963                 -0.533814  ...   
4                   0.530764                  1.243531  ...   

   -1.157572374481870670e+00  -3.169548928408494337e-01  \
0                   0.241794                  -0.937335   
1                   0.483655                   0.869513   
2                   0.385350                  -0.021720   
3                   0.661239                   0.329027   
4                   0.585614                  -0.317773   

   -1.041194344071040945e+00  -9.297458111228392719e-01  \
0                   0.288955                  -0.177624   
1                   0.881769                  -0.958674   
2                   0.863478                   0.256292   
3                  -0.757433                   0.082726   
4                  -0.637354                  -1.450445   

   -1.779338747140164045e+00  -1.429081038516362812e+00  \
0                   0.084456                   0.633965   
1                   1.880679                   0.608850   
2                   0.293780                   0.216499   
3                   0.267892                   0.282614   
4                  -1.656656                   0.924630   

   2.685259494797771929e+00  9.533337453891782554e-01  \
0                  1.052004                 -0.113860   
1                 -0.057137                 -0.950829   
2                 -0.637889                 -1.179666   
3                 -2.022049                 -2.201643   
4                 -1.116395                 -1.084778   

   -5.646716066842542014e-01  1.510000000000000000e+02  
0                   0.505908                      75.0  
1                  -0.594289                     141.0  
2                  -1.112571                     206.0  
3                   0.786962                     135.0  
4                   1.107924                      97.0  

[5 rows x 161 columns]

In [11]:
data['1.510000000000000000e+02'].value_counts(dropna=False)


200.0    6
72.0     6
90.0     5
71.0     5
178.0    5
        ..
40.0     1
161.0    1
45.0     1
308.0    1
317.0    1
Name: 1.510000000000000000e+02, Length: 214, dtype: int64

<hr>

## Question 2

<hr>

> Les variables explicatives sont-elles centrées ? Normalisées ? Qu’en est-il de la variable à expliquer ? Tracer un scatter plot de la base de données avec 4 covariables prises au hasard et la variable à expliquer (un scatterplot regroupe les graphes de chacune des variables en fonction de chacune des autres). Commenter les graphiques obtenus. (1.5 point)

<hr>

## Question 3

<hr>

> Echantillon d’apprentissage et de test. Créer 2 échantillons : un pour apprendre le modèle Xtrain, un pour tester le modèle Xtest. On mettra 25% de la base dans l’échantillon ’test’. Donner les tailles de chacun des 2 échantillons. On notera que le nouvel échantillon de covariables Xtrain n’est pas normalisé. Dans la suite, on fera donc bien attention à inclure l’intercept dans nos régressions. (1 point)

<hr>

## Question 4

<hr>

> Donner la matrice de covariance calculée sur Xtrain. Tracer le graphe de la décroissance des valeurs propres de la matrice de covariance (ou de corrélation). Expliquer pourquoi il est légitime de ne garder que les premières variables de l’ACP. On gardera 20 variables dans la suite. (1.5 point)


<hr>

## Question 5

<hr>

> Suivant les observations de la question (Q4), appliquer la méthode de "PCA before OLS" qui consiste à appliquer OLS avec Y et XtrainVp1:20q, où Vp1:20q contient les vecteurs propres (associés aux 20 plus grandes valeurs propres) de la matrice de covariance. Faire une régression linéaire (avec intercept), puis tracer les valeurs des coefficients (hors intercept). Sur un autre graphique, faire de même avec la méthode des moindres carrés classique. (1 point)Les variables explicatives sont-elles centrées ? Normalisées ? Qu’en est-il de la variable à expliquer ? Tracer un scatter plot de la base de données avec 4 covariables prises au hasard et la variable à expliquer (un scatterplot regroupe les graphes de chacune des variables en fonction de chacune des autres). Commenter les graphiques obtenus. (1.5 point)

<hr>

## Question 6

<hr>

> Donner les valeurs des intercepts pour les 2 régressions précédentes. Donner la valeur moyenne de la variable Y (sur le train set). Les intercepts des 2 questions sont-ils égaux ? Commenter. Uniquement pour cette question, centrer et réduire les variables après ACP (de petite dimension). Faire une régression avec ces variables et vérifier que l’intercept est bien égal à la moyenne de Y sur le train. (1 point)

<hr>

## Question 7

<hr>

> Pour les 2 méthodes (OLS et PCA before OLS) : Tracer les résidus de la prédiction sur l’échantillon test. Tracer leur densité (on pourra par exemple utiliser un histogramme). Calculer le coefficient de détermination sur l’échantillon test. Calculer le risque de prédiction sur l’échantillon test. (1.5 point)

<hr>

## Question 8

<hr>

> Coder la méthode de Forward variable sélection. On pourra utiliser la statistique du test de nullitédu coefficient (comme vu en cours). Pour l’instant, on ne met pas de critère d’arrêt sur la méthode, c’est-à-dire que l’on ajoute une variable à chaque étape jusqu’à retrouver la totalité des variables.Sachant que le jeu de données initiales ne contenait que les variables [22, 82, 23, 132, 154, 79, 115, 73, 122, 13], combien Forward en a-t-il retrouvé ? (3 points)

<hr>

## Question 9

<hr>

> Appliquer OLS sur les variables sélectionnées. Donner le risque de prédiction obtenu sur l’échantillon test et le comparer à ceux de OLS et PCA before OLS. (1.5 point)

<hr>

## Question 10

<hr>

> Afin de préparer la validation croisée, séparer l’échantillon train en 5 parties (appelées ”folds”) de façon aléatoire. On affichera les numéros d’échantillon sélectionnés dans chaque fold. (2 points)


<hr>

## Question 11

<hr>

> Appliquer la méthode de la régression Ridge. Pour le choix du paramètre de régularisation, on fera une validation croisée sur les ”folds” définis lors de la question précédente. A tour de rôle chacun des ”folds” servira à calculer le risque de prédiction alors que les autres seront utilisés pour estimer le modèle. On moyennera ensuite les 5 risques de prédictions. On donnera la courbe du risque de validation croisée en fonction du paramètre de régularisation (on veillera à bien choisir l’espace de définition pour le graphe). Donner le paramètre de régularisation optimal et la valeur du risque sur le test. (1.5 point)

<hr>

## Question 12

<hr>

> A l’aide de la fonction lassoCV de sklearn, choisir le paramètre de régularisation du LASSO. Donner le risque de prédiction associé. (1.5 point)

<hr>

## Question 13

<hr>

> Donner les variables selectionées par le Lasso. Combien y en a-t-il ? Appliquer la méthode OLS aux variables sélectionnées. Cette méthode est appelé Least-Square LASSO. (1 point)

<hr>

## Question 14

<hr>

> Les variables initiales du jeu de données sont [22, 82, 23, 132, 154, 79, 115, 73, 122, 13]. Combien le Lasso en a-t-il retrouvé ? (1 point)
